server/services/caldav/lib/calendar/calendar.syncUserCalendars.js

```diff
@@ -0,0 +1,189 @@
+  const CALDAV_HOME_URL = await this.gladys.variable.getValue('CALDAV_HOME_URL', this.serviceId, userId);
+  const CALDAV_USERNAME = await this.gladys.variable.getValue('CALDAV_USERNAME', this.serviceId, userId);
+  const CALDAV_PASSWORD = await this.gladys.variable.getValue('CALDAV_PASSWORD', this.serviceId, userId);
+
+  const xhr = new this.dav.transport.Basic(
+    new this.dav.Credentials({
+      username: CALDAV_USERNAME,
+      password: CALDAV_PASSWORD,
+    }),
+  );
+
+  const req = this.dav.request.propfind({
+    props: [
+      { name: 'calendar-description', namespace: this.dav.ns.CALDAV },
+      { name: 'calendar-timezone', namespace: this.dav.ns.CALDAV },
+      { name: 'displayname', namespace: this.dav.ns.DAV },
+      { name: 'getctag', namespace: this.dav.ns.CALENDAR_SERVER },
+      { name: 'resourcetype', namespace: this.dav.ns.DAV },
+      { name: 'supported-calendar-component-set', namespace: this.dav.ns.CALDAV },
+      { name: 'sync-token', namespace: this.dav.ns.DAV },
+    ],
+    depth: 1,
+  });
+
+  const listCalendars = await xhr.send(req, CALDAV_HOME_URL);
+
+  const davCalendars = listCalendars
+    .filter((res) => res.props.resourcetype.includes('calendar'))
+    .filter((res) => {
+      // We only want the calendar if it contains iCalendar objects.
+      const components = res.props.supportedCalendarComponentSet || [];
+      return components.reduce((hasObjs, component) => {
+        return hasObjs || ICAL_OBJS.has(component);
+      }, false);
+    })
+    .map((res) => {
+      logger.info(`CalDAV : Found calendar ${res.props.displayname}`);
+      return {
+        data: res,
+        description: res.props.calendarDescription,
+        timezone: res.props.calendarTimezone,
+        url: url.resolve(CALDAV_HOME_URL, res.href),
+        ctag: res.props.getctag,
+        displayName: res.props.displayname,
+        components: res.props.supportedCalendarComponentSet,
+        resourcetype: res.props.resourcetype,
+        syncToken: res.props.syncToken,
+      };
+    });
+
+  logger.info(`CalDAV : Found ${davCalendars.length} calendars.`);
+
+  const formatedCalendars = this.formatCalendars(davCalendars, userId);
+  const calendarsToUpdate = await Promise.all(
+    formatedCalendars.map(async (formatedCalendar) => {
+      const gladysCalendar = await this.gladys.calendar.get(userId, { externalId: formatedCalendar.external_id });
+      if (gladysCalendar.length === 0) {
+        const savedCalendar = await this.gladys.calendar.create(formatedCalendar);
+        delete savedCalendar.dataValues.sync_token;
+        return savedCalendar.dataValues;
+      }
+
+      if (formatedCalendar.ctag !== gladysCalendar[0].ctag) {
+        await this.gladys.calendar.update(gladysCalendar[0].selector, formatedCalendar);
+        return gladysCalendar[0];
+      }
+      return null;
+    }),
+  );
+
+  await Promise.map(
```

Comment:
Starting here, the function starts to be reaaaally big and have almost no comment.

It's hard to understand what you are doing !

Please add more comments, and maybe cut this function in smaller one, because here I'm lost :)

Reply:
Ok I'll work on it